packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -11,6 +11,33 @@ import 'package:flutter/foundation.dart';
 import 'text_editing.dart';
 import 'text_input.dart';
 
+/// {@template flutter.services.textFormatter.maxLengthEnforcement}
+/// How would the length limitation enforces in [LengthLimitingTextInputFormatter].
+///
+///  * For the behavior that enforce max length on the editing value entirely,
+/// choose [MaxLengthEnforcement.enforced], which is similar to Gboard and
+/// iOS native IME.
+///
+///  * For the behavior that exempt composing [TextEditingValue] while editing,
+/// choose [MaxLengthEnforcement.composingUnenforced]. A value without composing
+/// will still be truncated eventually.
+///
+///  * For the behavior that not enforcing the [TextEditingValue] but display
+/// a warning on the length counter, choose [MaxLengthEnforcement.warningOnly].
+/// {@endtemplate}
+enum MaxLengthEnforcement {
+  /// Enforce max length on the editing value entirely, which is similar to
+  /// Gboard and iOS native IME.
+  enforced,
+
+  /// Users can still input if the current value is composing while it reached
+  /// the max length limit. After composing ends, the value will be truncated.
+  composingUnenforced,
```

Comment:
IMO `allowComposingTextToFinish` still sounds a bit better. Any suggestions on how we should name these options @justinmc ?

Reply:
Yeah good point, I think I like:

```
enforced
allowComposingTextToFinish
none
```